src/Compilers/Core/Portable/InternalUtilities/StackGuard.cs

```diff
@@ -29,5 +30,27 @@ public static void EnsureSufficientExecutionStack(int recursionDepth)
                 RuntimeHelpers.EnsureSufficientExecutionStack();
             }
         }
+
+        internal static bool TryEnsureSufficientExecutionStack(int recursionDepth, bool throwOnFailure)
+        {
+            try
+            {
+                EnsureSufficientExecutionStack(recursionDepth);
+                return true;
+            }
+            catch (InsufficientExecutionStackException) when (!throwOnFailure)
+            {
+                return false;
+            }
+        }
+
+        internal static TResult ExecuteOnNewExecutionStack<TResult>(Func<TResult> execute)
+        {
+            var task = Task.Run(() => execute());
+            // Wait on the task without inlining the task on this thread.
+            Task.WhenAny(task).Wait();
+            // Return result, propagating any exception.
+            return task.GetAwaiter().GetResult();
```

Comment:
>GetResult [](start = 37, length = 9)

Before we make a decision about going ahead with this change, I think we should do some testing around exception handling and confirm what the actual behavior is. Here are the scenarios that come to mind:
- Would try catch (with and without filters) around the ```ExecuteOnNewExecutionStack``` method work exactly like if the code was executed on the same thread? If there are differences, what are they?
- Would debugging behavior be affected? I.e. would it change what debugger considers an unhandled exception?
- Would the stack on the original thread be easily discoverable/walkable in VS?
- Would quality of "Watson" dumps be affected in any way? For example, by unwinding the stack from the initial place, source of exception. Would the stack on the original thread be easily discoverable/walkable in VS?

Reply:
Unhandled exceptions on the child thread will be caught by the `Task` and re-thrown on the parent thread, so `try catch` blocks on the parent thread, with and without exception filters, should behave similarly to the single thread case today.

In the debugger, an exception that would be considered an unhandled exception today that is thrown on the child thread will be treated as an unhandled exception on the parent thread. If the debugger is set to break when an exception is thrown, the debugger will break when the exception is thrown on the child thread.

For crash dumps, the `FatalError.Report(e)` exception filter in `ExecuteOnNewExecutionStack()` will ensure the process crashes on the child thread rather than the parent thread.

The stacks for parent and child threads would be distinct and there would be no particular support for navigating to the parent stack when viewing a child stack.

Reply:
@AlekseyTs pointed out a behavior difference with exception filters: An exception filter on the parent thread will be executed after any `finally` blocks on the child thread.